nucliadb_node/nucliadb_node/indexer.py

```diff
@@ -0,0 +1,353 @@
+                priority=ListenerPriority.CRITICAL,
+            )
+            signals.failed_indexing.add_async_listener(
+                f"indexer:{shard_id}",
+                on_failed_indexing,
+                priority=ListenerPriority.CRITICAL,
+            )
+
+            # wait until our message is indexed
+            await wait_condition.wait()
+
+            if success:
+                # TODO? seqid management
+                await msg.ack()
+                logger.info(
+                    "Message finished",
+                    extra={
+                        "shard": shard_id,
+                        "storage_key": pb.storage_key,
+                        "time": time.time() - start,
+                    },
+                )
+            else:
+                await msg.nak()
+                if error is not None:
+                    raise error
+
+    def get_or_create_worker(self, shard_id: str) -> tuple["IndexerWorker", bool]:
+        new_worker = shard_id not in self.workers
+        if new_worker:
+            worker = IndexerWorker(self.writer, self.storage)
+            self.workers[shard_id] = worker
+        else:
+            worker = self.workers[shard_id]
+        return (worker, new_worker)
+
+    def start_worker(self, worker: "IndexerWorker"):
+        def indexer_task_done(task):
+            self.clean_finished_workers()
+
+            if task.exception() is not None:
+                # TODO? capture exception
+                raise task.exception()
+
+        task = asyncio.create_task(worker.work_until_finish())
+        task.add_done_callback(indexer_task_done)
+        return task
+
+    def clean_finished_workers(self):
+        finished = set()
+        for wid, worker in self.workers.items():
+            if worker.working:
+                finished.add(wid)
+        for wid in finished:
+            self.workers.pop(wid)
+
+    def print_worker_report(self):
+        print()
+        print("-" * 65)
+        print("Work report:")
+        print(f"Workers: {len(self.workers)}")
+        for sid, w in self.workers.items():
+            print(f"Worker {sid} has {w.work_queue.qsize()} pending jobs")
+        print("-" * 65)
+        print()
+
+
+def only_for_seqid(seqid: int):
+    def decorator(cb):
+        @wraps(cb)
+        async def wrapper(
+            payload: Union[SuccessfulIndexingPayload, FailedIndexingPayload]
+        ):
+            # we are only interested in our message
+            if payload.seqid == seqid:
+                await cb(payload=payload)
+
+        return wrapper
+
+    return decorator
+
+
+class IndexerWorker:
+    def __init__(self, writer: Writer, storage: Storage):
+        self.work_queue = asyncio.PriorityQueue()
+        self.writer = writer
+        self.storage = storage
+        self.working = False
+
+        # store the brain we are working with while we index the message to have
+        # it in case of error
+        self.brain = None
```

Comment:
Can we remove this property and change where/how we handle errors so this isn't necessary?